x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/indices/recovery/SearchableSnapshotRecoveryState.java

```diff
@@ -0,0 +1,115 @@
+        if (getStage() == Stage.DONE) {
+            return this;
+        }
+
+        // Pre-warm is still running, hold the state transition
+        // until the pre-warm process finishes
+        if (preWarmFinished == false && stage == Stage.DONE) {
+            return this;
+        }
+
+        return super.setStage(stage);
+    }
+
+    public synchronized void preWarmFinished() {
+        // For small shards it's possible that the
+        // cache is pre-warmed before the stage has transitioned
+        // to FINALIZE, so the transition to the final state is delayed until
+        // the recovery process catches up.
+        if (getStage() == Stage.FINALIZE) {
+            super.setStage(Stage.DONE);
+        }
+
+        SearchableSnapshotRecoveryState.Index index = (Index) getIndex();
+        index.stopTimer();
+        preWarmFinished = true;
+    }
+
+    public synchronized void ignoreFile(String name) {
+        SearchableSnapshotRecoveryState.Index index = (Index) getIndex();
+        index.addFileToIgnore(name);
+    }
+
+    private static final class Index extends RecoveryState.Index {
+        // We ignore the files that won't be part of the pre-warming
+        // phase since the information for those files won't be
+        // updated and marking them as reused might be confusing,
+        // as they are fetched on-demand from the underlying repository.
+        private final Set<String> filesToIgnore = new HashSet<>();
+
+        private Index() {
+            super(new SearchableSnapshotRecoveryFilesDetails());
+            // We start loading data just at the beginning
+            super.start();
+        }
+
+        private synchronized void addFileToIgnore(String name) {
+            filesToIgnore.add(name);
+        }
+
+        @Override
+        public synchronized void addFileDetail(String name, long length, boolean reused) {
+            if (filesToIgnore.contains(name)) {
+                return;
+            }
+
+            super.addFileDetail(name, length, reused);
+        }
+
+        // We have to bypass all the calls to the timer
+        @Override
+        public synchronized void start() {}
+
+        @Override
+        public synchronized void stop() {}
+
+        @Override
+        public synchronized void reset() {}
+
+        private synchronized void stopTimer() {
+            super.stop();
+        }
+    }
+
+    private static class SearchableSnapshotRecoveryFilesDetails extends RecoveryFilesDetails {
+        @Override
+        public void addFileDetails(String name, long length, boolean reused) {
+            // We allow reporting the same file details multiple times as we populate the file
+            // details before the recovery is executed and therefore we ignore the rest
+            // of the calls for the same files.
+            fileDetails.computeIfAbsent(name, n -> new FileDetail(name, length, reused));
```

Comment:
This means that the reused flag is not overridden, but based on what it was initially set?

Reply:
That's correct.

Reply:
This sounds wrong then, as we would have reused == true for all files (as we initially fake that all files are present), which is wrong as we can only reuse what we actually locally have.

Reply:
Those file details are populated first on `SearchableSnapshotDirectory#prewarmCache`, and that process is triggered before the recovery process starts. Meaning that they should contain the correct information.

Reply:
Can we assert something to that effect?

Reply:
I'm thinking that maybe the intent will be clearer if we set a flag after the information is populated in `SearchableSnapshotDirectory#prewarmCache` and we bypass the calls to this method once that flag is true, wdyt?

Reply:
yes

Reply:
Finally I've kept the same approach as it's possible that during peer recoveries an additional `segments_n` file is sent over as after restore the index is marked with a new history uuid generating a new `segments_` file that's unknown to the target peer. So I've added a stronger assertion to ensure that if the file was already added it has the same length.